### PR TITLE
feat(arrows.ts): change reduce to a loop and extract to a function + add tests

### PR DESCRIPTION
**Pull request by VIKTORVAV99**
## Issue

The reduce function has a lot of overhead compared to simple loops which causes it to be slower.

## Description

This PR changes the use of the reduce methods into a single reusable function that implements a function instead to speed things up a bit.

This PR also adds appropriate testing of the new function.

As an added benefit of the reusable code this PR reduces our bundle size a tiny bit.

### Double check

- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

___

## **Type**
enhancement, tests

___

## **Description**
- Introduced a new `filterExchanges` function in `arrows.ts` to efficiently filter exchanges based on zone and country exclusions.
- Utilized `Set` for faster exclusion checks within the `filterExchanges` function.
- Replaced old filtering logic in `useExchangeArrowsData` with calls to `filterExchanges`, enhancing code readability and performance.
- Added a suite of tests in `arrows.test.ts` to ensure the `filterExchanges` function works as expected across various scenarios, including error handling.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>arrows.ts</strong><dd><code>Refactor Exchange Filtering and Improve Performance</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

web/src/hooks/arrows.ts
<li>Extracted logic for filtering exchanges based on zone and country view <br>into a new <code>filterExchanges</code> function.<br> <li> Replaced previous filtering logic with calls to the new <br><code>filterExchanges</code> function.<br> <li> Utilized <code>Set</code> for exclusion lists to improve lookup performance.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6437/files#diff-eca68a824b3bab38bc89b0be58e84cb0adc73083dc456edbf8dafcabe6624b7f">+30/-12</a>&nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>arrows.test.ts</strong><dd><code>Add Tests for Exchange Filtering Function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

web/src/hooks/arrows.test.ts
<li>Added comprehensive tests for the new <code>filterExchanges</code> function.<br> <li> Tests cover normal cases, edge cases, and error handling.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6437/files#diff-0f0029bb7afd704196d32625b81bb0e8b03fab9a093ef088781bbc859674ac27">+126/-0</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: web/src/hooks/arrows.test.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  },
+  exchange4: {
+    '2024-01-31T18:00:00Z': { co2intensity: 199.73, netFlow: -18 },
+    '2024-01-31T19:00:00Z': { co2intensity: 195.2, netFlow: -18 },
+  },
+  exchange5: {
+    '2024-01-31T20:00:00Z': { co2intensity: 188.22, netFlow: -11 },
+    '2024-01-31T21:00:00Z': { co2intensity: 178.86, netFlow: -50 },
+  },
+};
```
Comment (VIKTORVAV99) [inline web/src/hooks/arrows.test.ts:30]: FYI I used GitHub Copilot to generate this test data after giving it a real snippet. Really handy.

File: web/src/hooks/arrows.ts [hunk 2]
```diff
@@ -16,6 +16,30 @@ import exchangesToExclude from '../../config/excluded_aggregated_exchanges.json'
...
+  // Loop through the exchanges and assign them to the correct result object
+  for (const key of Object.keys(exchanges)) {
+    if (exclusionSetZones.has(key)) {
+      resultCountries[key] = exchanges[key];
+      continue;
+    }
+    if (exclusionSetCountries.has(key)) {
+      resultZones[key] = exchanges[key];
+    }
+  }
```
Comment (tonypls) [inline web/src/hooks/arrows.ts:39]:
```suggestion
  for (const [key, value] of Object.entries(exchanges)) {
    if (exclusionSetZones.has(key)) {
      resultCountries[key] = value;
      continue;
    }
    if (exclusionSetCountries.has(key)) {
      resultZones[key] = value;
    }
  }
```

Maybe a slight performance increase?

**Review by tonypls — APPROVED** [review]
This is super cool, did you notice any difference or try profiling?

**Comment by VIKTORVAV99** [comment]
> This is super cool, did you notice any difference or try profiling?

I didn't try profiling (maybe I should though) but just based on what I know about callbacks, reduce and the fact that it's run once instead of twice should have a significant impact.

But honestly I mostly split it out into a different function so we could test it more easily.

**Comment by VIKTORVAV99** [comment]
Object entries are faster according to this little benchmarking suit https://www.measurethat.net/Benchmarks/Show/3686/0/object-values-objectentries-vs-objectkeys-vs-objectkeys so switching to that.
